whistleblower/whistleblower_bot/process_event.py

```diff
@@ -0,0 +1,191 @@
+
+
+def message_for_app(event):
+    event_type = event.get("type")
+    return event_type == "app_mention" or (
+        event_type == "message" and event.get("channel_type") == "im" and _check_bot_im(event)
+    )
+
+
+def _get_bot_user_id():
+    if utils.redis_client.exists("bot_user_id"):
+        bot_user_id = utils.redis_client.get("bot_user_id").decode("utf-8")
+    else:
+        response = utils.web_api_client.auth_test()
+        if not response.get("ok", False):
+            logger.error(
+                "Auth test for bot user failed. Returned error: {}".format(response.get("error"))
+            )
+            return None
+        bot_user_id = response.get("user_id")
+        utils.redis_client.set("bot_user_id", bot_user_id)
+
+    return bot_user_id
+
+
+def _check_bot_im(event):
+    response = utils.web_api_client.conversations_members(channel=event.get("channel"))
+    if not response.get("ok", False):
+        logger.error(
+            "Failed to get conversation members for IM chat. Returned error: {}".format(
+                response["error"]
+            )
+        )
+
+    return _get_bot_user_id() in response.get("members")
+
+
+def process_app_message(event):
+    message = event.get("text", "")
+
+    phrases, channels = utils.split_message(message)
+    if phrases and channels:
+        _update_user_config(event, phrases, channels)
+    else:
+        if "purge config" in message:
+            _purge_user_config(event)
+        elif "get config" in message:
+            _get_user_config(event)
+        else:
+            _chat_post_message(event.get("channel"), utils.HELP_MESSAGE)
+
+
+def _get_user_config(event):
+    user = event.get("user")
+    if not user:
+        logger.warning("User is not present in incomming event, could not retrieve config")
+        return
+    if utils.redis_client.exists(user):
+        user_conf = json.loads(utils.redis_client.get(user))
+        _ack_config_get_msg(user_conf["phrases"], user_conf["channels"], event.get("channel"))
+    else:
+        _chat_post_message(event.get("channel"), message="You don't have any configuration!")
+
+
+def _ack_config_get_msg(phrases, channels, chan_post, base_msg=None):
+    msg = utils.DISPLAY_CONF_MESSAGE.format(
+        "\n".join(phrase for phrase in phrases), " ".join(channels)
+    )
+    if base_msg:
+        msg = base_msg + msg
+    _chat_post_message(chan_post, msg)
+
+
+def _purge_user_config(event):
+    user_id = event.get("user")
+
+    if utils.redis_client.exists(user_id):
+        utils.redis_client.delete(user_id)
+
+    for chan_id in utils.redis_client.scan_iter(match="C*"):
+        chan_conf = json.loads(utils.redis_client.get(chan_id))
+        phrases_users = chan_conf.get("phrases", {})
+
+        stale_phrases = []
+        for phrase, users in phrases_users.items():
+            if user_id in users:
+                users.remove(user_id)
+                if not users:
+                    stale_phrases.append(phrase)
```

Comment:
How would you refactor this code to avoid 4-level nesting of code blocks?

Reply:
@mmatviyiv brb at this, as soon as i am near my computer :)